### DISCUSSION TITLE
Annotations and Redactions

### DISCUSSION THREAD

**Discussion by mwade-noetic** [py-pdf/pypdf#1548] [Q&A]
Hi,

   Great to see this pure python project alive and moving forward.  Just wondering if you have support for doing Redactions as well as Annotations.  I did a quick survey of the documentation but there is no mention of Redactions.  I did a quick search and I see support for READING a lot of different annotation types (25 I think), but it wasn't clear to me that you could actually write most of them.  

The redaction support is obviously not simple when you have to consider layers of objects (text overlaying an image for example). 

Thanks,

Michael Wade

**Comment by pubpub-zz** [comment]
Yes you can read (https://pypdf.readthedocs.io/en/3.2.0/user/reading-pdf-annotations.html) and write (https://pypdf.readthedocs.io/en/3.2.0/user/adding-pdf-annotations.html) annotations 😉😊

**Reply by mwade-noetic** [reply]
Thanks, As I noted, I was aware of the ability to read and write (some annotations).  However, there are a LOT of annotation types that I do not see supported on the write side (Highlight's for example).  In particular I am looking to see if they support doing Redactions.  Typically that is a type of Annotation, and then there is some separate functionality to burn them in.  One method is to create the redaction annotation and then you can load the PDF in Acrobat Pro and burn in the redactions.   But it doesn't appear that I can write that type of annotation object.

**Reply by pubpub-zz** [reply]
I see what you mean.
You should have a look at the Pdf 1.7 reference §8.4.5 for details about the different annotations. The highlight actually is just asking for the  vertices of the quadrilateral.  The only issue is to be able to calculate these coordinates. The solution I would propose - for the moment - would be to use a visitor function with the extract_text function.
